Add the Auto Import feature for Code Completion

junichi11 opened the pull request:
- https://github.com/apache/netbeans/issues/6947
- Add `Auto Import` as an option for code completion
- Add `Prefer Import` and `Don't Import` options for global namespace items(`Don't Import` is enabled by default)
- Add `File Scope`(unchecked by default) and `Namespace Scope`(checked by default) options
  - `File Scope` means a php file without a namespace name (e.g. `<html><p><?php echo \NS\something(); ?></p></html>`)
- Don't add a use statement if use list has the same name item(Instead, the result of "Smart" or "Unqualified" CC is used)
- Add unit tests


Note: A use statement may not be added to an expected position if the existing use list is not sorted(ignore cases)

### Options

![nb-php-auto-import-options-03](https://github.com/apache/netbeans/assets/738383/bc67a8a7-87bc-4418-bda6-146cec3a7316)

### Single Use

![nb-php-auto-import-02](https://github.com/apache/netbeans/assets/738383/19f9f8b6-b0de-47e7-9e1a-6c58d9d6d263)

### Group Use

![nb-php-auto-import-with-group-use](https://github.com/apache/netbeans/assets/738383/b8243a69-b469-4f2e-bdeb-a3f819e1eafe)

junichi11 commented on php/php.editor/src/org/netbeans/modules/php/editor/completion/PHPCompletionItem.java:
```diff
@@ -359,12 +368,99 @@ public String getInsertPrefix() {
+                String fqName = ifq.getFullyQualifiedName().toString().substring(CodeUtils.NS_SEPARATOR.length());
+                String name = getName();
+                String useType = getUseType();
+                String aliasName = CodeUtils.EMPTY_STRING;
```
This is not used at the moment. However, it would be useful if someone requests importing as an alias(when the use list has the same name) in the future.

tmysik replied:
I see.

junichi11 commented:
@tmysik Could you please have a look at this when you have time? (Huge change, sorry...)

junichi11 commented:
Dev build to test this: https://github.com/apache/netbeans/suites/20437665071/artifacts/1219370144

tmysik commented on php/php.editor/src/org/netbeans/modules/php/editor/codegen/AutoImport.java:
```diff
@@ -0,0 +1,910 @@
+    public static final String USE_TYPE = "type"; // NOI18N
+    public static final String USE_FUNCTION = "function"; // NOI18N
+    public static final String USE_CONST = "const"; // NOI18N
+    private static final Logger LOGGER = Logger.getLogger(AutoImport.class.getName());
```
Nitpick: Don't we put `LOGGER` always as the first field in a class?

junichi11 replied:
Will do that :) Thanks!

junichi11 replied:
Fixed.

tmysik commented on php/php.editor/src/org/netbeans/modules/php/editor/completion/PHPCompletionItem.java:
```diff
@@ -129,16 +131,22 @@ public abstract class PHPCompletionItem implements CompletionProposal {
+
+    final CompletionRequest request;
+    private final ElementHandle element;
+    private QualifiedNameKind generateAs;
```
Can we put non-final fields below the final ones?

junichi11 replied:
Will do that. Thanks!

junichi11 replied:
Fixed.

tmysik reviewed (approved):
Well, overall, it looks good to me. Of course, to review all the details is difficult since it is a big change :) But we have many tests for it, this is great :+1:

junichi11 commented:
@tmysik Thank you for your time and your review! I'll improve this a bit because there are a few feedbacks: https://github.com/apache/netbeans/issues/6947#issuecomment-1926982191

junichi11 commented:
Changed the options a bit.
![nb-php-auto-import-options-01](https://github.com/apache/netbeans/assets/738383/f34a9878-adc2-4578-9b2d-eec682a3c8b1)

Fixed field orders and improve the insert position.

https://github.com/apache/netbeans/compare/9a40e281060cc08ce3dc9e9e4e3b51c5170389bc..b31c4fea571c5fcfc1496efc8d2eb91c8bcf0f6c

Thanks!

tmysik commented:
Cool!
